Add a parallel COPY for large objects using the MultiPart Upload API

copiousfreetime opened the pull request:
Based off of the s3-mp-upload.py script, this script does a parallel copy of an S3 object between S3 buckets.

copiousfreetime committed:
add s3-mp-copy.py

mumrah commented:
Neat! I didn't know the MultiPart supported COPY like this. I'll check this out when I get some time, thanks for the contribution.

-David

copiousfreetime commented:
It does require a newer released version of boto. This is the only way to copy objects greater than 5GB between buckets without downloading and uploading them via an ec2 instance. I just had a good test with chunk size  of 100M and 40 parallel processes.

`INFO:s3-mp-copy:Finished copying 171394.32M in 385.18s (444.97Mbps)`

copiousfreetime committed:
fix copying when the size is smaller than multipart threshold

copiousfreetime committed:
Fix off by 1 error in chunk copying.

The last chunk of the S3 object that was getting copied was not getting copied
if the S3 object was greater than 5GB.

mumrah commented:
@copiousfreetime I'm going to go ahead and merge this. I'll be sure to assign any bugs reported to you :)
